File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        res=[]
        def binarys(res,n):
            l=0
            r=len(res)-1
            while l<=r:
                mid=(l+r)//2
                if res[mid]==n:
                    return mid
                elif res[mid]>n:
                    r=mid-1
                else:
                    l=mid+1
            return l
        for n in nums:
            if not res or res[-1]<n:
                res.append(n)
            else:
                id=binarys(res,n)
                res[id]=n
        return len(res)

File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        int n=nums.length;
        int dp[][]=new int[n][n+1];
        for (int i = 0; i < n; i++) {
            Arrays.fill(dp[i], -1);
        }
        return f(0,-1,n,nums,dp);
    }
    public int f(int ind,int prev_ind,int n,int nums[],int dp[][]){
        if(ind==n) return 0;
        if(dp[ind][prev_ind+1]!=-1) return dp[ind][prev_ind+1];
        int len=0+f(ind+1,prev_ind,n,nums,dp);
        if(prev_ind==-1||nums[ind]>nums[prev_ind])
        len=Math.max(len,1+f(ind+1,ind,n,nums,dp));
        return dp[ind][prev_ind+1]=len;

    }
}